### ISSUE TITLE
BACKEND

### ISSUE THREAD

**Issue by aman77singh** [aman77singh/Aman-Singh#2]
# Cell 1: Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pennylane as qml
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize

# Cell 2: Load the Dataset
data = pd.read_csv(r'C:\Users\aman\Desktop\hackathon\archive\creditcard.csv')

# Cell 3: Select Features and Scale Them
features = data.drop(['Class'], axis=1)  # Exclude the target variable
target = data['Class']

# Use only the first 10,000 rows
features = features.iloc[:10000]
target = target.iloc[:10000]

# Scale the features
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

# Select only the first 27 features
features_subset = features_scaled[:, :27]  # Limit to 27 features

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features_subset, target, test_size=0.2, random_state=42)

# Cell 4: Set Up the Quantum Device
n_features = features_subset.shape[1]
n_qubits = n_features  # Using one qubit per feature

# Create a device
dev = qml.device('qiskit.aer', wires=n_qubits)

# Define a quantum circuit
@qml.qnode(dev)
def circuit(weights):
    # Encode data into quantum state
    for i in range(n_qubits):
        qml.RX(X_train[0, i], wires=i)  # Encoding using RX gates
    
    # Apply a simple circuit
    for i in range(n_qubits - 1):
        qml.CNOT(wires=[i, i + 1])
    
    # Measurement
    return qml.probs(wires=range(n_qubits))

# Cell 5: Define the Softmax Function and Cost Function
def softmax(x):
    e_x = np.exp(x - np.max(x))  # Shift for numerical stability
    return e_x / e_x.sum(axis=0)

def cost(weights):
    predictions = circuit(weights)
    probabilities = softmax(np.abs(predictions)**2)  # Convert amplitudes to probabilities

    # Use a small epsilon to prevent log(0)
    epsilon = 1e-6
    probabilities[1] = np.clip(probabilities[1], epsilon, 1 - epsilon)
    probabilities[0] = np.clip(probabilities[0], epsilon, 1 - epsilon)

    return -np.sum(y_train * np.log(probabilities[1]) + (1 - y_train) * np.log(probabilities[0]))

# Cell 6: Train the Model
# Initialize weights randomly
initial_weights = np.random.rand(n_qubits)

# Define a callback function to monitor progress
def callback(weights):
    print(f"Current weights: {weights}")
    print(f"Current cost: {cost(weights)}")

# Optimize weights using a classical optimizer
opt_result = minimize(cost, initial_weights, method='BFGS', callback=callback)
optimal_weights = opt_result.x

# Cell 7: Making Predictions
def predict(weights, data):
    predictions = []
    for i in range(data.shape[0]):
        prediction = circuit(data[i])
        probabilities = softmax(np.abs(prediction)**2)  # Convert amplitudes to probabilities
        predictions.append(probabilities[1])  # Probability of class 1
    return np.array(predictions)

# Get predictions
y_pred_probs = predict(optimal_weights, X_test)

# Convert probabilities to binary predictions
y_pred = (y_pred_probs > 0.5).astype(int)

# Cell 8: Evaluate the Model
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

**Comment by aman77singh** [comment]
# Cell 1: Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pennylane as qml
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize

# Cell 2: Load the Dataset
data = pd.read_csv(r'C:\Users\aman\Desktop\hackathon\archive\creditcard.csv')

# Cell 3: Select Features and Scale Them
features = data.drop(['Class'], axis=1)  # Exclude the target variable
target = data['Class']

# Use only the first 10,000 rows
features = features.iloc[:10000]
target = target.iloc[:10000]

# Scale the features
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

# Select only the first 27 features
features_subset = features_scaled[:, :27]  # Limit to 27 features

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features_subset, target, test_size=0.2, random_state=42)

# Cell 4: Set Up the Quantum Device
n_features = features_subset.shape[1]
n_qubits = n_features  # Using one qubit per feature

# Create a device
dev = qml.device('qiskit.aer', wires=n_qubits)

# Define a quantum circuit
@qml.qnode(dev)
def circuit(weights):
    # Encode data into quantum state
    for i in range(n_qubits):
        qml.RX(X_train[0, i], wires=i)  # Encoding using RX gates
    
    # Apply a simple circuit
    for i in range(n_qubits - 1):
        qml.CNOT(wires=[i, i + 1])
    
    # Measurement
    return qml.probs(wires=range(n_qubits))

# Cell 5: Define the Softmax Function and Cost Function
def softmax(x):
    e_x = np.exp(x - np.max(x))  # Shift for numerical stability
    return e_x / e_x.sum(axis=0)

def cost(weights):
    predictions = circuit(weights)
    probabilities = softmax(np.abs(predictions)**2)  # Convert amplitudes to probabilities

    # Use a small epsilon to prevent log(0)
    epsilon = 1e-6
    probabilities[1] = np.clip(probabilities[1], epsilon, 1 - epsilon)
    probabilities[0] = np.clip(probabilities[0], epsilon, 1 - epsilon)

    return -np.sum(y_train * np.log(probabilities[1]) + (1 - y_train) * np.log(probabilities[0]))

# Cell 6: Train the Model
# Initialize weights randomly
initial_weights = np.random.rand(n_qubits)

# Define a callback function to monitor progress
def callback(weights):
    print(f"Current weights: {weights}")
    print(f"Current cost: {cost(weights)}")

# Optimize weights using a classical optimizer
opt_result = minimize(cost, initial_weights, method='BFGS', callback=callback)
optimal_weights = opt_result.x

# Cell 7: Making Predictions
def predict(weights, data):
    predictions = []
    for i in range(data.shape[0]):
        prediction = circuit(data[i])
        probabilities = softmax(np.abs(prediction)**2)  # Convert amplitudes to probabilities
        predictions.append(probabilities[1])  # Probability of class 1
    return np.array(predictions)

# Get predictions
y_pred_probs = predict(optimal_weights, X_test)

# Convert probabilities to binary predictions
y_pred = (y_pred_probs > 0.5).astype(int)

# Cell 8: Evaluate the Model
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

**Comment by aman77singh** [comment]
# Cell 1: Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pennylane as qml
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize

# Cell 2: Load the Dataset
data = pd.read_csv(r'C:\Users\aman\Desktop\hackathon\archive\creditcard.csv')

# Cell 3: Select Features and Scale Them
features = data.drop(['Class'], axis=1)  # Exclude the target variable
target = data['Class']

# Use only the first 10,000 rows
features = features.iloc[:10000]
target = target.iloc[:10000]

# Scale the features
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

# Select only the first 27 features
features_subset = features_scaled[:, :27]  # Limit to 27 features

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features_subset, target, test_size=0.2, random_state=42)

# Cell 4: Set Up the Quantum Device
n_features = features_subset.shape[1]
n_qubits = n_features  # Using one qubit per feature

# Create a device
dev = qml.device('qiskit.aer', wires=n_qubits)

# Define a quantum circuit
@qml.qnode(dev)
def circuit(weights):
    # Encode data into quantum state
    for i in range(n_qubits):
        qml.RX(X_train[0, i], wires=i)  # Encoding using RX gates
    
    # Apply a simple circuit
    for i in range(n_qubits - 1):
        qml.CNOT(wires=[i, i + 1])
    
    # Measurement
    return qml.probs(wires=range(n_qubits))

# Cell 5: Define the Softmax Function and Cost Function
def softmax(x):
    e_x = np.exp(x - np.max(x))  # Shift for numerical stability
    return e_x / e_x.sum(axis=0)

def cost(weights):
    predictions = circuit(weights)
    probabilities = softmax(np.abs(predictions)**2)  # Convert amplitudes to probabilities

    # Use a small epsilon to prevent log(0)
    epsilon = 1e-6
    probabilities[1] = np.clip(probabilities[1], epsilon, 1 - epsilon)
    probabilities[0] = np.clip(probabilities[0], epsilon, 1 - epsilon)

    return -np.sum(y_train * np.log(probabilities[1]) + (1 - y_train) * np.log(probabilities[0]))

# Cell 6: Train the Model
# Initialize weights randomly
initial_weights = np.random.rand(n_qubits)

# Define a callback function to monitor progress
def callback(weights):
    print(f"Current weights: {weights}")
    print(f"Current cost: {cost(weights)}")

# Optimize weights using a classical optimizer
opt_result = minimize(cost, initial_weights, method='BFGS', callback=callback)
optimal_weights = opt_result.x

# Cell 7: Making Predictions
def predict(weights, data):
    predictions = []
    for i in range(data.shape[0]):
        prediction = circuit(data[i])
        probabilities = softmax(np.abs(prediction)**2)  # Convert amplitudes to probabilities
        predictions.append(probabilities[1])  # Probability of class 1
    return np.array(predictions)

# Get predictions
y_pred_probs = predict(optimal_weights, X_test)

# Convert probabilities to binary predictions
y_pred = (y_pred_probs > 0.5).astype(int)

# Cell 8: Evaluate the Model
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

**Comment by aman77singh** [comment]
# Cell 1: Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pennylane as qml
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize

# Cell 2: Load the Dataset
data = pd.read_csv(r'C:\Users\aman\Desktop\hackathon\archive\creditcard.csv')

# Cell 3: Select Features and Scale Them
features = data.drop(['Class'], axis=1)  # Exclude the target variable
target = data['Class']

# Use only the first 10,000 rows
features = features.iloc[:10000]
target = target.iloc[:10000]

# Scale the features
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

# Select only the first 27 features
features_subset = features_scaled[:, :27]  # Limit to 27 features

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features_subset, target, test_size=0.2, random_state=42)

# Cell 4: Set Up the Quantum Device
n_features = features_subset.shape[1]
n_qubits = n_features  # Using one qubit per feature

# Create a device
dev = qml.device('qiskit.aer', wires=n_qubits)

# Define a quantum circuit
@qml.qnode(dev)
def circuit(weights):
    # Encode data into quantum state
    for i in range(n_qubits):
        qml.RX(X_train[0, i], wires=i)  # Encoding using RX gates
    
    # Apply a simple circuit
    for i in range(n_qubits - 1):
        qml.CNOT(wires=[i, i + 1])
    
    # Measurement
    return qml.probs(wires=range(n_qubits))

# Cell 5: Define the Softmax Function and Cost Function
def softmax(x):
    e_x = np.exp(x - np.max(x))  # Shift for numerical stability
    return e_x / e_x.sum(axis=0)

def cost(weights):
    predictions = circuit(weights)
    probabilities = softmax(np.abs(predictions)**2)  # Convert amplitudes to probabilities

    # Use a small epsilon to prevent log(0)
    epsilon = 1e-6
    probabilities[1] = np.clip(probabilities[1], epsilon, 1 - epsilon)
    probabilities[0] = np.clip(probabilities[0], epsilon, 1 - epsilon)

    return -np.sum(y_train * np.log(probabilities[1]) + (1 - y_train) * np.log(probabilities[0]))

# Cell 6: Train the Model
# Initialize weights randomly
initial_weights = np.random.rand(n_qubits)

# Define a callback function to monitor progress
def callback(weights):
    print(f"Current weights: {weights}")
    print(f"Current cost: {cost(weights)}")

# Optimize weights using a classical optimizer
opt_result = minimize(cost, initial_weights, method='BFGS', callback=callback)
optimal_weights = opt_result.x

# Cell 7: Making Predictions
def predict(weights, data):
    predictions = []
    for i in range(data.shape[0]):
        prediction = circuit(data[i])
        probabilities = softmax(np.abs(prediction)**2)  # Convert amplitudes to probabilities
        predictions.append(probabilities[1])  # Probability of class 1
    return np.array(predictions)

# Get predictions
y_pred_probs = predict(optimal_weights, X_test)

# Convert probabilities to binary predictions
y_pred = (y_pred_probs > 0.5).astype(int)

# Cell 8: Evaluate the Model
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

**Comment by aman77singh** [comment]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pennylane as qml
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize

# Cell 2: Load the Dataset
data = pd.read_csv(r'C:\Users\aman\Desktop\hackathon\archive\creditcard.csv')

# Cell 3: Select Features and Scale Them
features = data.drop(['Class'], axis=1)  # Exclude the target variable
target = data['Class']

# Use only the first 10,000 rows
features = features.iloc[:10000]
target = target.iloc[:10000]

# Scale the features
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

# Select only the first 27 features
features_subset = features_scaled[:, :27]  # Limit to 27 features

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features_subset, target, test_size=0.2, random_state=42)

# Cell 4: Set Up the Quantum Device
n_features = features_subset.shape[1]
n_qubits = n_features  # Using one qubit per feature

# Create a device
dev = qml.device('qiskit.aer', wires=n_qubits)

# Define a quantum circuit
@qml.qnode(dev)
def circuit(weights):
    # Encode data into quantum state
    for i in range(n_qubits):
        qml.RX(X_train[0, i], wires=i)  # Encoding using RX gates
    
    # Apply a simple circuit
    for i in range(n_qubits - 1):
        qml.CNOT(wires=[i, i + 1])
    
    # Measurement
    return qml.probs(wires=range(n_qubits))

# Cell 5: Define the Softmax Function and Cost Function
def softmax(x):
    e_x = np.exp(x - np.max(x))  # Shift for numerical stability
    return e_x / e_x.sum(axis=0)

def cost(weights):
    predictions = circuit(weights)
    probabilities = softmax(np.abs(predictions)**2)  # Convert amplitudes to probabilities

    # Use a small epsilon to prevent log(0)
    epsilon = 1e-6
    probabilities[1] = np.clip(probabilities[1], epsilon, 1 - epsilon)
    probabilities[0] = np.clip(probabilities[0], epsilon, 1 - epsilon)

    return -np.sum(y_train * np.log(probabilities[1]) + (1 - y_train) * np.log(probabilities[0]))

# Cell 6: Train the Model
# Initialize weights randomly
initial_weights = np.random.rand(n_qubits)

# Define a callback function to monitor progress
def callback(weights):
    print(f"Current weights: {weights}")
    print(f"Current cost: {cost(weights)}")

# Optimize weights using a classical optimizer
opt_result = minimize(cost, initial_weights, method='BFGS', callback=callback)
optimal_weights = opt_result.x

# Cell 7: Making Predictions
def predict(weights, data):
    predictions = []
    for i in range(data.shape[0]):
        prediction = circuit(data[i])
        probabilities = softmax(np.abs(prediction)**2)  # Convert amplitudes to probabilities
        predictions.append(probabilities[1])  # Probability of class 1
    return np.array(predictions)

# Get predictions
y_pred_probs = predict(optimal_weights, X_test)

# Convert probabilities to binary predictions
y_pred = (y_pred_probs > 0.5).astype(int)

# Cell 8: Evaluate the Model
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))
